models/gif_psc_exp.cpp

```diff
@@ -0,0 +1,486 @@
+    >= 0 ); // since t_ref_ >= 0, this can only fail in error
+
+  // initializing internal state
+  if ( !S_.sfa_stc_initialized_ )
+  {
+    V_.P_sfa_.clear();
+    V_.P_stc_.clear();
+    S_.sfa_elems_.clear();
+    S_.stc_elems_.clear();
+
+    for ( size_t i = 0; i < P_.tau_sfa_.size(); i++ )
+    {
+      V_.P_sfa_.push_back( std::exp( -h / P_.tau_sfa_[ i ] ) );
+      S_.sfa_elems_.push_back( 0.0 );
+    }
+
+    for ( size_t i = 0; i < P_.tau_stc_.size(); i++ )
+    {
+      V_.P_stc_.push_back( std::exp( -h / P_.tau_stc_[ i ] ) );
+      S_.stc_elems_.push_back( 0.0 );
+    }
+
+    S_.sfa_stc_initialized_ = true;
+  }
+}
+
+/* ----------------------------------------------------------------
+ * Update and spike handling functions
+ */
+
+void
+nest::gif_psc_exp::update( Time const& origin,
+  const long_t from,
+  const long_t to )
+{
+
+  assert(
+    to >= 0 && ( delay ) from < kernel().connection_manager.get_min_delay() );
+  assert( from < to );
+
+  double_t q_temp_;
+
+  for ( long_t lag = from; lag < to; ++lag )
+  {
+
+    // exponential decaying stc and sfa elements
+    q_temp_ = 0;
+    for ( size_t i = 0; i < S_.stc_elems_.size(); i++ )
+    {
+
+      q_temp_ += S_.stc_elems_[ i ];
+
+      S_.stc_elems_[ i ] = V_.P_stc_[ i ] * S_.stc_elems_[ i ];
+    }
+
+    S_.stc_ = q_temp_;
+
+    q_temp_ = 0;
+    for ( size_t i = 0; i < S_.sfa_elems_.size(); i++ )
+    {
+
+      q_temp_ += S_.sfa_elems_[ i ];
+
+      S_.sfa_elems_[ i ] = V_.P_sfa_[ i ] * S_.sfa_elems_[ i ];
+    }
+
+    S_.sfa_ = q_temp_ + P_.V_T_star_;
+
+    // exponential decaying PSCs
+    S_.i_syn_ex_ *= V_.P11ex_;
+    S_.i_syn_in_ *= V_.P11in_;
+
+    S_.i_syn_ex_ += B_.spikes_ex_.get_value( lag );
+    S_.i_syn_in_ += B_.spikes_in_.get_value( lag );
+
+    if ( S_.r_ref_ == 0 ) // neuron not refractory, so evolve V
+    {
+
+      if ( S_.add_stc_sfa_ )
+      {
+
+        S_.add_stc_sfa_ = false;
+
+
+        q_temp_ = 0;
+
+        for ( size_t i = 0; i < S_.stc_elems_.size(); i++ )
+        {
+          S_.stc_elems_[ i ] += P_.q_stc_[ i ];
+
+          q_temp_ += P_.q_stc_[ i ];
+        }
+
+        S_.stc_ += q_temp_;
+
+
+        q_temp_ = 0;
+
+        for ( size_t i = 0; i < S_.sfa_elems_.size(); i++ )
+        {
+
+          S_.sfa_elems_[ i ] += P_.q_sfa_[ i ];
+
+          q_temp_ += P_.q_sfa_[ i ];
+        }
+
+        S_.sfa_ += q_temp_;
+      }
+
+      S_.y3_ = V_.P30_ * ( S_.y0_ + P_.I_e_ - S_.stc_ ) + V_.P33_ * S_.y3_
+        + V_.P31_ * P_.E_L_ + S_.i_syn_ex_ * V_.P21ex_
+        + S_.i_syn_in_ * V_.P21in_;
+
+      double_t lambda =
+        P_.lambda_0_ * std::exp( ( S_.y3_ - S_.sfa_ ) / P_.Delta_V_ );
+
+
+      if ( lambda > 0.0 )
+      {
+
+        // Draw random number and compare to prob to have a spike
+        // hazard function is computed by 1 - exp(- lambda * dt)
+        if ( V_.rng_->drand()
+          < -numerics::expm1( -lambda * Time::get_resolution().get_ms() ) )
+        {
+          S_.add_stc_sfa_ = true;
```

Comment:
As mentioned above, my understanding of the Mensi et al paper (especially eqs 3 and 5) is that the jumps in stc and sfa should occur here.